Added HAProxy support

LightningReflex opened the pull request:
Cool I added support for HAProxy, so if the server has HAProxy enabled, it can still be pinged using this program by setting "HAProxy: true"

LightningReflex committed:
Added HAProxy support

woodendoors7 commented:
I will have to fix some things about that PR, we stopped using buffers (and I am on a holiday), though interestingly I haven't heard of HAproxy yet, I assume you need to register in the proxy first by sending those headers? Have you tested it already? Also it seems like it only handles ipv4 rn, and does Minecraft send a HAproxy header as well, is it custom in its code? + what is the purpose of the header? Can't find much about it when I google, it looks like haproxy can be used as a normal tcp proxy?

LightningReflex commented:
Hey, thanks for the response, by default Minecraft doesn't support the HAProxy header, but paper's added support for it since 2022.

HAProxy is essentially a load-balancing system, but it causes the user's IP-address to get lost, and replaced with the load-balancer's IP. That's where an HAProxyMessage comes in, this packet contains the original IP address of the user, and allows you to preserve the IP address when HAProxy is used.

Now, when you enable the HAProxy option on paper (config/paper-global.yml ->proxies.proxy-protocol: true), this causes any normal connection (including status pings) directly to the server without the HAProxyMessage to stop functioning (unhandled, no response).

What I added in the PR made it so the status ping also sends a HAProxyMessage which allows the status ping to be handled correctly by paper.

I have tested this on paper/purpur, but not on vanilla (as HAProxyMessages aren't supported on it), so I'm unsure what it would produce if you attempted to send an HAProxyMessage to it.

If you need a better explanation, feel free to dm me on discord, I've already friended you.

woodendoors7 commented:
Hi, I forgot to respond, but as this is nothing that minecraft supports, nor is it needed, I will not merge this. Plus, the implementation is not the greatest - splits IPv4s manually with dots, uses buffer when we don't use it anymore, but mainly, it only supports a hardcoded local IPv4 address and a hardcoded port.

The best way to do it is to ping your intended server through the proxy - there should be no problem with that, and by doing that, you can also catch a problem with the proxy sooner than you would when pinging it directly.

I think this would be better handled with some API in the library that allows you to change the TCP transaction as it goes on, not many purpose-specific mods, however, this is not a TCP library, and that way it'd become something akin to that - I think the best course of action is forking it and making your little version tweaked to your needs, if you still need to bypass the proxy - and that's why stuff is open source.

I appreciate for taking the time to make this PR, and feel free to submit any quality PRs in the future :3
